// awsencryptionsdk-js/lib/datamodel.js
var Model = {
    newAWSEncodedObject:  {
        header : {
            Version : "",
            Type : "",
            AlgorithmID : "",
            MessageID : "",
            AADLength : "",
            AADBytes : "",
            AAD : "",
            EncryptedDataKeyCount : "",
            EncryptedDataKeys : [
                {
                    KeyProviderIDLength : "",
                    KeyProviderID : "",
                    KeyProviderInformationLength : "",
                    KeyProviderInformation : "",
                    EncryptedDataKeyLength : "",
                    EncryptedDataKey : ""
                }
            ],
            ContentType : "",
            Reserved : "",
            IVLength : "",
            FrameLength : "",
            HeaderAuthentication : {
                IV: "",
                AuthenticationTag: ""
            }
        },
        body : {
            frames : [],
            lastFrame: {}
        },
        footer : {}
    } 
}

module.exports = Model;
